Convert Frontend Test Runner from Minitest to RSpec

KludgeKML opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

**NB: Currently failing checks because it's expecting a minitest check - as long as that's the only failure it is safe to merge but we will need to update the settings**

## What

Replace Frontend's existing minitest test suite with the equivalent in RSpec.

## Why

RSpec is our standard, and we want to merge some smaller apps (using RSpec test suites) into Frontend. We have decided it's better to rewrite the larger app to conform to our standard than move the smaller app in the other direction.

[Trello card](https://trello.com/c/Xw0X4LDe/159-consolidate-frontend-feedback-test-frameworks), [Jira issue PNP-8578](https://gov-uk.atlassian.net/browse/PNP-8578)

## How

We do the bulk of the conversion using [a fork](https://github.com/KludgeKML/minitest_to_rspec/tree/bulk-convert) of the [minitest_to_rspec](https://github.com/jaredbeck/minitest_to_rspec) gem; then tidy up the code with rubocop before manually tweaking the remaining tests that don't run. Finally we compare coverage to ensure that we have the same between the versions.

### Coverage

Line coverage is slightly improved (97.8% -> 97.94%)

### Style decisions I have made (and can unmake!)

In tidying up the autogenerated specs, I've made a bunch of style decisions that aren't really covered by rubocop but which I've tried to maintain to keep the new files consistent with other spec tests (we don't have a perfect consistency in other apps, but there's a rough consensus). I'm not married to these decisions, so feel free to comment on them and we can update the files:

- requires at the top of _spec files don't have parentheses around their parameters
- describe/context/it blocks don't have parentheses around their parameter
- describe blocks at the top of a file are namespaced explicitly (ie RSpec.describe instead of just describe)
- includes at the top of a describe block don't have parentheses around their parameters
- expect _does_ have parentheses (if not a block), and so do the matches, but `#to` and `#not_to` don't.
- There are blank links between contexts and its.
- Inside an it block, I try to leave a blank line between setup and the first expectation
- I use `#describe` for the top block in a file, and for any block that is a named method, `#context` otherwise.
- I don't have separate assertions for "check for a status of redirect" and "check the current location" - expect(response).to redirect_to(<url>) covers both of these tests.

### Test Type Conversion

The current idiom is to prefer system spec over feature specs, so all feature tests have been converted to system specs, and request specs over controller specs, so all controller tests have been converted to request specs. Note that in a few cases this has meant a slight adjustment to the tests (some previous controller tests were not suitable for request specs, so have been moved into the system specs for the related feature, and some tests have been simplified as above).

In addition, I've moved tests into directories that reflect their test types, allowing us to avoid explicit test types with teh `config.infer_spec_type_from_file_location!` RSpec directive - so some of the previous unit tests are moved out into /models and /helpers. Ideally no test types should be added explicitly. The sole exception to this is components, since by convention we tend to put those in /components, and they need to be a mixture of view types (the actual component tests) and default type (the all_components code).

### Decisions I have avoided making

I haven't closely examined each test to determine whether it's actually worthwhile as a test. Some of the reviewers pointed out tests that are not great, I've dealt with them. But don't take this PR as an endorsement of the quality of individual tests.

KludgeKML committed:
Fix bool method returning non-bool values

KludgeKML committed:
Update personalisation method calls to new names

leenagupte commented on app/helpers/components/calendar_helper.rb:
```diff
@@ -0,0 +1,7 @@
+module Components
+  module CalendarHelper
+    def time_tag_safe(date)
```
Ah, it's not a new helper method, it was originally here: https://github.com/alphagov/frontend/blob/9f4ce5c8d6d7db29bcbf71382da9efe24d8af190/app/helpers/calendar_helper.rb#L8

Does that mean there are two modules called `CalendarHelper` now?

leenagupte replied:
I think there's a stepping missing to remove the `time_tag_safe` method from the original `CalendarHelper` module.

KludgeKML replied:
This is a problem that I identified and fixed in this branch, but then [got fixed in the main branch](https://github.com/alphagov/frontend/pull/4099) as well. I will rebase and remove my changes.

leenagupte commented on app/controllers/account_home_controller.rb:
```diff
@@ -2,6 +2,6 @@ class AccountHomeController < ApplicationController
 
   def show
-    redirect_with_analytics GovukPersonalisation::Urls.your_account
+    redirect_with_analytics GovukPersonalisation::Urls.one_login_your_services
```
Nice!

KludgeKML replied:
(I'll mark all the positive comments as resolved).

leenagupte commented on Gemfile:
```diff
@@ -41,9 +41,8 @@ end
 group :test do
   gem "govuk_schemas"
   gem "i18n-coverage"
-  gem "mocha"
```
Did you mean to remove "mocha" in this commit as well?

KludgeKML replied:
I did (we don't need it any more)

leenagupte replied:
I see that Mocha tests etc are removed later in the PR, but at this point, they're are still there. Ideally the gem would be removed after, or at the same time that the use of Mocha is removed.

leenagupte commented on spec/support/govuk_spec.rb:
```diff
@@ -0,0 +1 @@
+GovukTest.configure
```
I think should this be in `spec_helper`.

KludgeKML replied:
Usage does say that it can be here: https://github.com/alphagov/govuk_test?tab=readme-ov-file#usage, but probably doesn't add anything. I'll move it to rails_helper (the spec_helper for this app is still the generated one).

leenagupte commented on spec/support/content_schema_helpers.rb:
```diff
@@ -1,20 +1,20 @@
-    def valid_payload_for(format_name)
-      ValidSchemaMatcher.new(format_name)
-    end
+# module Mocha
```
Why did this need to be disabled and why only temporarily?

KludgeKML replied:
It was copied across, but was blocking any test running. I've actually removed it entirely now, it's not really possible to do the test in the same way, and it didn't really add much.

leenagupte commented on spec/components/calendar_spec.rb:
```diff
@@ -0,0 +1,24 @@
+require("rails_helper")
+describe("CalendarComponent", type: :view) do
```
I think we should decide at this point what format the view tests will take. i.e. should they be [feature tests](https://github.com/rspec/rspec-rails/blob/main/features/view_specs/view_spec.feature)?

KludgeKML replied:
I've deliberately put aside the question of changing test types (see above), but I think view specs are appropriate for the components, at the very least (https://rspec.info/features/6-0/rspec-rails/view-specs/view-spec/) - they're for testing templates independently of a controller, so it seems fine for component tests (where we want to test that the component is being rendered but could be in multiple controllers). Existing specs for components  (outside of the gem, which uses minitest) seem to use view specs (https://github.com/alphagov/collections/blob/main/spec/components/document_navigation_list_spec.rb)

leenagupte commented on spec/feature/account_home_spec.rb:
```diff
@@ -0,0 +1,22 @@
+    stub_content_store_has_item("/")
+  end
+
+  context("/account/home") do
```
It's strange that `GOVUK_PERSONALISATION_YOUR_ACCOUNT_URI` is not being modified here. 🤔

KludgeKML replied:
Unless I've misunderstood something it doesn't really need to be in this test, because the default values for those methods are fine.

leenagupte commented on spec/feature/account_home_spec.rb:
```diff
@@ -0,0 +1,22 @@
+  end
+
+  context("/account/home") do
+    it("redirect users to One Login's Your Services page") do
```
I'm guessing the script just converted `should` to `it` to get the test expectations working and that the test names will need to be manually modified by a human later?

KludgeKML replied:
Yes, in fact I've done that now as part of a general clean-up (I'll add details of my arbitrary style decisions in the description so they can be critiqued).

leenagupte reviewed (commented):
This is looking good. And the tests are running 🎉 

A couple of things I'm not sure about is the nesting of the components helper modules. Other thoughts / comments are inline.

KludgeKML committed:
Update CI, Rake to use RSpec

hannako commented on spec/rails_helper.rb:
```diff
@@ -0,0 +1,29 @@
+  config.include FeatureHelpers, type: :feature
+
+  config.include Helpers::Components, type: :view
+end
```
I think we don't need this file right? If we put all the config into the new spec_helper file (added in the **Install RSpec and init** commit) then I think we can lose all of the `require "rails_helper"` statements at the top of all the specs

KludgeKML replied:
Hmm - the general rule AFAIK is that you want separate spec_helper and rails_helper files in case you are developing locally and want to run a subset of your tests (in which case you only include rails_helper stuff in the files that actually need rails_helper). Maybe we don't run subsets of the test suite locally enough to justify that, though? It makes no difference to CI.

hannako replied:
There's no rails helper in collections. What else uses RSpec?

KludgeKML replied:
Hmm, I was going to say imminence has a rails_helper but no spec_helper, but I did that conversion from minitest too, so that could just be my mistake.

KludgeKML replied:
I've converted it here to spec_helper and removed the includes. Seems like a good idea, and easy enough to put them back if necessary.

hannako reviewed (commented):
Nice work!

KludgeKML commented on spec/support/content_schema_helpers.rb:
```diff
@@ -0,0 +1,20 @@
+module Mocha
```
Ah, I thought I'd got rid of it (in fact I think I have, I just accidentally got rid of it in a later commit and missed that it should be here). Thanks, I'll sort that out!

KludgeKML replied:
Done!

KludgeKML commented on spec/support/content_store_helpers.rb:
```diff
@@ -0,0 +1,75 @@
+    content_item
+  end
+
+  def stub_content_store_has_item_tagged_to_taxon(base_path:, payload:)
```
Ah yes, thanks! (I've checked and removed all the now unused methods from this file)

KludgeKML commented on spec/support/content_store_helpers.rb:
```diff
@@ -0,0 +1,75 @@
+    GdsApi.stubs(:content_store).returns(content_store)
+  end
+
+  def content_store_has_archived_page(slug)
```
Thanks! (- see above)

leenagupte commented on spec/support/content_store_helpers.rb:
```diff
@@ -0,0 +1,55 @@
+    content_item
+  end
+
+  def content_store_has_page(slug, schema: "generic", is_tagged_to_taxon: false)
```
I'm not sure why this method is needed. Can the tests not call `content_store_has_random_item` with the specific base_path they require?

leenagupte commented on spec/support/election_helpers.rb:
It'd be really helpful to split this commit up so that the helpers are added along with the tests using them, so it's easy to see how they're used. Especially for specific helpers like these ElectionHelpers.

KludgeKML replied:
Done (marking resolved)

leenagupte commented on app/models/calendar.rb:
```diff
@@ -1,5 +1,5 @@
 class Calendar
-  REPOSITORY_PATH = Rails.env.test? ? "test/fixtures" : "lib/data"
+  REPOSITORY_PATH = Rails.env.test? ? "spec/fixtures" : "lib/data"
```
Oh no. I didn't realise this clause was in the model 😬.  A follow-up task should be to stub `REPOSITORY_PATH` in the tests. 😬

leenagupte commented on spec/components/all_components_spec.rb:
```diff
@@ -0,0 +1,48 @@
+    template = filename.split("/").last
+    component_name = template.sub("_", "").sub(".html", "").sub(".erb", "").gsub("-", "_")
+
+    context(component_name) do
```
This is a very boring comment, but I've seen some people use `context` and some people use `describe`. I think we should standardise on one. Or an at least decide  when one should be used over the other.

KludgeKML replied:
I've detailed my use of describe vs context in the PR description. (Basically, describe when it's top level or the comment is just a method name, context otherwise, which I think makes most sense semantically)

leenagupte replied:
So I think then that this should be a `describe` rather than a `context`?

I've generally thought that `describe` is used for "things" e.g. method names, routes, etc and `context` is used to describe the circumstances of the test, e.g. a set of tests for postcode outcomes, could have a context of "in England".

But this is testing a components, which I think is a "thing" not a "circumstance".

KludgeKML replied:
Oh, I see. We're agreeing on style, you're just pointing out there's a mistake here. Yes, it should be a describe.

leenagupte commented on spec/components/calendar_spec.rb:
```diff
@@ -0,0 +1,25 @@
+describe "CalendarComponent", type: :view do
```
Another boring comment, but I think we should use `RSpec.describe` consistently. Just using `describe` is valid too, but I think most of our test suites use the `RSpec` namespace.

leenagupte commented on spec/components/all_components_spec.rb:
Could this commit be split up a bit? I think by type of test, e.g. component tests, or if the spec file is really large it should have its own commit. Also, if the spec has it's own helper I think they should be committed together so it's easy to see what goes with what.
I'd also delete the relevant minitest test in the same commit so it's easy to compare what's being replaced.

KludgeKML replied:
Done (going to mark this as resolved)

leenagupte reviewed (commented):
I think this change is on the right track, and it's great that the test coverage hasn't decreased.

This is a pretty huge PR, which is not surprising considering the scale of the change, however that means though I've had a bit of trouble following the changes. Some of the commits are very large and it's hard to compare what's being replaced.  It would great if the commits could be broken down into more atomic commits, so the related changes can be seen together.

I've also added some inline comments regarding styling and which conventions to follow.

KludgeKML committed:
Install RSpec and init

KludgeKML committed:
Convert help_cookies integration tests

KludgeKML committed:
Convert help integration tests

KludgeKML committed:
Convert homepage integration tests

KludgeKML committed:
Convert icalendar integration tests

- update to use CalenderHelpers instead
  of test code in Calendar model

KludgeKML committed:
Convert json integration tests

- update to use CalendarHelpers rather
  than test code in Calendar class.

KludgeKML committed:
Convert place integration tests

KludgeKML committed:
Convert sessions integration tests

- reengineer these tests to avoid the weird
  try/rescue block

KludgeKML committed:
Convert travel_advice_atom integration tests

KludgeKML committed:
Convert travel_advice integration tests

KludgeKML committed:
Convert find_local_council_controller functional tests

KludgeKML committed:
Convert help_controller functional tests

KludgeKML committed:
Convert licence_transaction_controller functional tests

KludgeKML committed:
Convert place_controller functional tests

KludgeKML committed:
Convert placeholder_controller functional tests

KludgeKML committed:
Convert random_controller functional tests

KludgeKML committed:
Convert roadmap_controller functional tests

KludgeKML committed:
Convert session_controller functional tests

KludgeKML committed:
Convert simple_smart_answers_controller functional tests

KludgeKML committed:
Convert transaction_controller functional tests

KludgeKML committed:
Convert travel_advice_controller functional tests

KludgeKML committed:
Convert simple_smart_answers routing tests

- replace utility function from content_store_helper
  only called in one place

KludgeKML committed:
Convert all service unit tests

- Remove content_schema_helpers, and reference in
  the CalendarPublisher spec. For the test we're
  making here, it's a lighter touch to remove the
  dependency on Mocha and just match the actual
  example rather than validating that it matches
  a schema.

KludgeKML committed:
Move pact helper

KludgeKML committed:
Convert account_home integration tests

- Integration tests from here on become system tests
- We create a mostly empty content_store_helper, we'll
   add the methods we use as we come to them. For this
   test it's just there so that it can include the GDS
   content store helper methods.

KludgeKML committed:
Convert component tests

- convert component_helpers
- add component helper to view specs
- fix unsafe yaml load

KludgeKML committed:
Convert error_handling integration tests

- move content_store_throws_exception_for in content
  store helper (first use)

KludgeKML committed:
Convert electoral_controller functional tests

KludgeKML committed:
Convert transaction integration tests

- First use of content_store_has_example_item, so bring it into content store helpers
- Only user of schema_org_helpers, but that's quite a big set of information, so copy across the file rather than move it into the using test.

KludgeKML committed:
Convert calendar_controllers functional tests

- functional tests are converted into request
  tests from here on.
- names are changed to remove the Controller
  suffix.
- include content store helpers into request tests
  as they almost all use them.
- move honours_content_store_ttl over
- split out some tests to routing tests - one
  test is clearly a routine test, one needs both
  a routing and request test to cover its behaviour.

KludgeKML committed:
Convert local_transaction_controller functional tests

KludgeKML committed:
Convert application_helper unit tests

- Move all helpers into /helpers so that
  we can infer test type.

KludgeKML committed:
Convert error_items_helper unit tests

KludgeKML committed:
Convert phone_number_helper unit tests

KludgeKML committed:
Convert election_postcode model unit tests

- Model tests moved out of the units directory
  and into /models where the infer_type setting
  will set them up correctly.

KludgeKML committed:
Convert location_error model unit tests

KludgeKML committed:
Convert uprn model unit tests

KludgeKML committed:
Convert calendar_content_item presenter unit tests

KludgeKML committed:
Convert content_item presenter unit tests

KludgeKML committed:
Convert electoral presenter unit tests

- Move the fixture file used by this test

KludgeKML committed:
Convert faq presenter unit tests

- Update to use CalendarHelpers rather
  than test code in Calendar class.

KludgeKML committed:
Convert licence_details presenter unit tests

KludgeKML committed:
Convert local_authority presenter unit tests

KludgeKML committed:
Convert local_transaction presenter unit tests

KludgeKML committed:
Convert place presenter unit tests

KludgeKML committed:
Convert simple_smart_answer presenter unit tests

KludgeKML committed:
Convert transaction presenter unit tests

KludgeKML committed:
Convert travel_advice_index presenter unit tests

KludgeKML committed:
Convert bank_holiday integration tests

- create bank holiday helper with bank holiday table assertion
- create calender helper (used here and in later
  tests) to allow us to take test-specific code
  out of the Calendar model
- convert assert_bank_holiday_table to a custom matcher and
  move it into its own support/matcher file.
- move fixtures used for these tests (and others)

KludgeKML committed:
Convert when_do_the_clocks_change integration tests

- Update to use CalendarHelpers rather
  than test code in Calendar class.
- Move fixtures

KludgeKML committed:
Convert recruitment_banner_helper unit tests

- Move the fixture file used by this test

KludgeKML committed:
Convert api_error_routing_constraint unit tests

KludgeKML committed:
Convert bank_holiday_generator unit tests

- Move fixture file used by theses tests

KludgeKML committed:
Convert format_routing_constraint unit tests

KludgeKML committed:
Convert ics_renderer unit tests

KludgeKML committed:
Convert locales_validation unit tests

- Also suppress the "good job!" message which
  messes up the neat RSpec dots.

KludgeKML committed:
Convert postcode sanitizer unit tests

KludgeKML committed:
Convert calendar unit tests

- Update Calendar spec to use
  CalendarHelper rather than test
  code in Calendar class.
- Move fixture used in this test

KludgeKML committed:
Remove obsolete minitest files

- delete unused fixture
- delete _helper files from test
- remove test-related gems
- remove content_store_helpers from test/support,
  the only remaining functions in it were never
  used.

KludgeKML committed:
Remove obsolete code

- Neither of these methods are referenced in the app.

leenagupte reviewed (commented):
I got as far as reviewing commit 99d5ff7ace1149b8a4e31bc6077014b849b8a017 "Convert homepage integration tests".

As agreed IRL, this is a partial review only up to and including that commit.

The change is looking good so far. Thanks for breaking up the commits, it has made it much easier to review. 🎉 

One of the issues that I keep stumbling over is how the conversion tool as flattened all of the hashes and made them pretty unreadable. 

I've left some other comments inline.

KludgeKML commented on spec/components/all_components_spec.rb:
```diff
@@ -0,0 +1,48 @@
+        expect(File).to exist(css_file)
+      end
+
+      it "does not use `html_safe` in it's partial" do
```
Yeah, possibly! I actually have a bigger plan for this - this whole file (pretty much) should be a shared example that you can import from the publishing component gem (then you would have one test that says "do all the components have tests", and each individual component spec file could have "behaves like a component", which would cover all of these tests (and would potentially improve with each improvement in the gem)

KludgeKML replied:
(marking as resolved, will continue in govuk_components_gem)

KludgeKML commented on spec/components/all_components_spec.rb:
```diff
@@ -0,0 +1,48 @@
+RSpec.describe "AllComponents", type: :controller do
```
I refined my thinking on this a bit doing the components in government-frontend, so you're right it definitely shouldn't be a controller test. It might be that a view test is best, but I'll check whether taking the type out makes no difference, and if so I'll do that.

KludgeKML replied:
Done.

KludgeKML commented on spec/components/all_components_spec.rb:
```diff
@@ -0,0 +1,48 @@
+      end
+
+      it "has the correct documentation" do
+        yaml = YAML.unsafe_load_file(yaml_file)
```
Strong suspicion it's because whoever wrote the code just didn't know about permitting classes during a safe load (if you just change it to safe load it will fail, but if you change it to safe load and permit Time objects, it's okay. I've done that now)

KludgeKML replied:
(Marking as resolved)

KludgeKML commented on spec/spec_helper.rb:
```diff
@@ -17,6 +17,7 @@
 GovukTest.configure
 
 RSpec.configure do |config|
+  config.include ComponentHelpers, type: :feature
```
Nope (see commit message) - component helpers includes a lot of additional assertions used all over the feature tests. But I'll look into if there are more RSpec-y ways of doing them that might make this unnecessary.

KludgeKML replied:
We've got rid of feature tests, so I removed this.

hannako commented on spec/feature/help_spec.rb:
```diff
@@ -0,0 +1,14 @@
+    it "renders the help index page correctly" do
+      visit "/help"
+
+      expect(page).to have_css("h1", text: "Help using GOV.UK")
```
Minor: `expect(page).to have_title("Help using GOV.UK")`

KludgeKML replied:
Done. Thanks! that makes much more sense in the tests.

hannako commented on spec/feature/help_cookies_spec.rb:
```diff
@@ -0,0 +1,24 @@
+    it "renders the cookies setting page correctly" do
+      visit "/help/cookies"
+
+      within("#content") { expect(page).to have_css("h1", text: "Cookies on GOV.UK") }
```
Minor: `expect(page).to have_title("Cookies on GOV.UK")`

KludgeKML replied:
Done!

hannako commented on spec/feature/homepage_spec.rb:
```diff
@@ -0,0 +1,26 @@
+RSpec.describe "Homepage", type: :feature do
+  before { stub_content_store_has_item("/", schema: "special_route") }
```
We'll be adding a schema example for the homepage content item soon so can update this test then.

KludgeKML replied:
Cool! :)

hannako commented on spec/feature/icalendar_spec.rb:
```diff
@@ -0,0 +1,53 @@
+
+  context "getting ICS version" do
+    before do
+      allow(File).to receive(:mtime).with(Rails.root.join("REVISION")).and_return(Time.zone.parse("2012-10-17 01:00:00"))
```
Urgh what's this?! Maybe the spec pre-dates [TimeHelpers](https://api.rubyonrails.org/classes/ActiveSupport/Testing/TimeHelpers.html#method-i-travel_to). Can we use that instead?

hannako commented on spec/feature/json_spec.rb:
```diff
@@ -0,0 +1,38 @@
+
+    it "has the full calendar json view" do
+      visit "/bank-holidays.json"
+      expected = { "england-and-wales" => { "division" => "england-and-wales", "events" => [{ "date" => "2012-01-02", "notes" => "Substitute day", "title" => "New Year\u2019s Day", "bunting" => true }, { "date" => "2012-06-04", "notes" => "Substitute day", "title" => "Spring bank holiday", "bunting" => true }, { "date" => "2012-06-05", "notes" => "Extra bank holiday", "title" => "Queen\u2019s Dia ...
```
Argh my eyes! I can see Leena has commented on all these hashes already

hannako commented on spec/support/button_as_link_helpers.rb:
```diff
@@ -0,0 +1,23 @@
+module ButtonAsLinkHelpers
+  RSpec::Matchers.define :have_button_as_link do |text, attrs = {}|
```
I would expect to find a custom matcher in the RSpec config I think. i.e. in the spec_helper file. But that's just my opinion

KludgeKML replied:
I think it makes more sense in this case to have it as a separate file (as with the other one), but open to a vote?

KludgeKML replied:
Done

hannako commented on spec/feature/place_spec.rb:
```diff
@@ -0,0 +1,240 @@
+  end
+
+  context "when visiting the start page" do
+    before { visit "/passport-interview-office" }
```
I think the spacing has been lost in the conversion in this file

KludgeKML replied:
Fixed

hannako commented on spec/feature/place_spec.rb:
```diff
@@ -0,0 +1,240 @@
+      end
+      within("#content") do
+        within(".gem-c-title") do
+          expect(page).to have_css("h1", text: "Find a passport interview office")
```
`.to have_title`

KludgeKML replied:
Done

hannako commented on spec/feature/sessions_spec.rb:
```diff
@@ -1,59 +1,52 @@
-      assert response.location = GovukPersonalisation::Urls.manage
+      visit(new_govuk_session_callback_path(code: "code", state: "state"))
+
+      expect(page.current_url).to eq("#{GovukPersonalisation::Urls.one_login_your_services}/")
```
Similar to a previous comment, we're not checking for the redirect anymore. Is that just because it seemed overkill?

KludgeKML replied:
Checking for the current path in a system test is fine (it will fail if there isn't a redirect, so it does both jobs)

hannako commented on spec/support/feature_helpers.rb:
```diff
@@ -7,4 +7,8 @@ def assert_current_url(path_with_query, options = {})
     end
   end
+
+  def assert_page_has_content(text)
```
Not a blocker, but I feel like we don't need this custom helper method. There are a LOT of places where we're not using this method and doing `expect page.to have_content`

KludgeKML replied:
Done! (this whole file is removed now)

hannako reviewed (commented):
~~I'm up to commit 23 "Convert simple_smart_answers integration tests"~~

Done!

hannako commented on spec/unit/format_routing_constraint_spec.rb:
```diff
@@ -1,56 +1,58 @@
+  include ContentStoreHelpers
 
-class FormatRoutingConstraintTest < ActiveSupport::TestCase
   context "#matches?" do
```
Should the first `context` be a `describe`

KludgeKML replied:
Done!

hannako commented on spec/unit/locales_validation_spec.rb:
```diff
@@ -0,0 +1,7 @@
+  it "validates all locale files" do
+    checker = RailsTranslationManager::LocaleChecker.new("config/locales/*.yml")
+
+    expect(checker.validate_locales).to(be_truthy)
```
the bracket looks strange, should it be `expect(checker.validate_locales).to be_truthy`

KludgeKML replied:
Done!

KludgeKML commented on spec/support/bank_holiday_helpers.rb:
```diff
@@ -0,0 +1,10 @@
+module BankHolidayHelpers
```
(Gonna mark resolved)

KludgeKML commented on spec/requests/csv_preview_spec.rb:
```diff
@@ -0,0 +1,52 @@
+require "gds_api/test_helpers/asset_manager"
+require "gds_api/test_helpers/content_store"
+
+RSpec.describe "CSV Preview", type: :request do
```
(going to mark as resolved)

KludgeKML commented on spec/requests/csv_preview_spec.rb:
```diff
@@ -0,0 +1,52 @@
+require "gds_api/test_helpers/asset_manager"
+require "gds_api/test_helpers/content_store"
+
+RSpec.describe "CSV Preview", type: :request do
```
So, now that we're actually converting these to proper system tests, we can have that one test back in the same file (making this split and the problems outlined above go away) :)

KludgeKML commented on spec/requests/csv_preview_spec.rb:
```diff
@@ -0,0 +1,52 @@
+
+RSpec.describe "CSV Preview", type: :request do
+  include GdsApi::TestHelpers::AssetManager
+  include GdsApi::TestHelpers::ContentStore
```
Done!

KludgeKML commented on spec/requests/csv_preview_spec.rb:
```diff
@@ -0,0 +1,52 @@
+      content_store_filename = "2023-01-11_filename"
+      path_from_filename = ->(base_name) { "media/#{asset_manager_id}/#{base_name}.csv" }
+      setup_asset_manager(parent_document_url, asset_manager_id, asset_manager_filename)
+      setup_content_item(path_from_filename[content_store_filename], parent_document_base_path)
```
No longer being duplicated. :)

KludgeKML commented on spec/requests/csv_preview_spec.rb:
```diff
@@ -0,0 +1,52 @@
+
+  asset_manager_id = 4321
+  parent_document_base_path = "/government/important-guidance"
+  parent_document_url = "https://www.test.gov.uk#{parent_document_base_path}"
```
They're actually being used all over the file. I've converted them to lets, though.

KludgeKML commented on spec/requests/csv_preview_spec.rb:
```diff
@@ -0,0 +1,52 @@
+    csv_file = if malformed
+                 " \"Field 1\"\n"
+               else
+                 nonconvertible ? "\x8D\n" : generate_test_csv(51, 1010)
```
Not 100% sure what this comment was (I can't find the comment in the other test?)

KludgeKML replied:
Oh, I remember - it was about this being a weird construct with the if/else/trinary. I've updated to an if/elsif/else

KludgeKML committed:
Convert csv_preview integration tests

- take out the explicit inclusion of content
  store helper because that's now brought into
  all system tests.
- There's a single test that we want to make a
  request test because we are only interested in
  the location of the redirection, and setting up the
  scaffolding for this in a system test would
  be too awkward.
- As a result, we make a helper so that we're not
  repeating the complex setup code.

KludgeKML committed:
Convert electoral_look_up integration tests

- convert election_helpers (first use)
- move used fixture

KludgeKML commented on spec/support/election_helpers.rb:
```diff
@@ -17,7 +17,7 @@ def stub_api_address_lookup(uprn, response: "{}", status: 200)
 
   def api_response
-    path = Rails.root.join("test/fixtures/electoral-result.json")
+    path = Rails.root.join("spec/fixtures/electoral-result.json")
```
Now moved in this commit.

KludgeKML committed:
Convert find_local_council integration tests

- local_links_manager_helpers only has one method
  only used in one place, so move it to there.
- First appearance of content_store_has_random_item,
  so bring it into content_store_helper
- bring in location_helpers (first use)

KludgeKML committed:
Convert local_transactions integration tests

KludgeKML committed:
Convert licence_transaction integration tests

- First  use of assert_current_url - but we don't want to use that
  when we can use expect(page.current_path).to eq, so remove it
  from the test helper at this point.
- create button_as_link custom matcher in support/matchers
  replacing old custom assertion

KludgeKML committed:
Convert simple_smart_answers integration tests

- first use of assert_page_has_content, we will replace it
  with expect(page).to have_content from now on, so remove
  from the test helper

KludgeKML commented on spec/components/calendar_spec.rb:
```diff
@@ -0,0 +1,25 @@
+  end
+
+  it "renders the component with correct data passed" do
+    render_component(title: "Marvel films", year: "2008 onwards", headings: [{ text: "Date" }, { text: "Day of the week" }, { text: "Film" }], events: [{ title: "Iron Man", date: Date.parse("2-5-2008"), notes: "first film in the MCU" }, { title: "The Incredible Hulk", date: Date.parse("13-6-2008"), notes: "" }])
```
Done

KludgeKML commented on spec/feature/account_home_spec.rb:
```diff
@@ -0,0 +1,20 @@
+    stub_content_store_has_item("/")
+  end
+
+  context "/account/home" do
```
In fact has become a system test.

KludgeKML commented on spec/feature/bank_holidays_spec.rb:
```diff
@@ -0,0 +1,206 @@
+RSpec.describe "BankHolidays", type: :feature do
```
I've fixed this now to make them system tests, with a simpler runner (rack_test)

KludgeKML commented on spec/feature/bank_holidays_spec.rb:
```diff
@@ -0,0 +1,206 @@
+        within(".govuk-tabs") do
+          within("#england-and-wales") do
+            expect(page).to have_link("Add bank holidays for England and Wales to your calendar", href: "/bank-holidays/england-and-wales.ics")
+            assert_bank_holiday_table(title: "Upcoming bank holidays in England and Wales", year: "2012", rows: [["Date", "Day of the week", "Bank holiday"], ["25 December", "Tuesday", "Christmas Day"], ["26 December", "Wednesday", "Boxing Day"]])
```
Done

KludgeKML commented on spec/feature/gwyliau_banc_spec.rb:
```diff
@@ -0,0 +1,126 @@
+        within(".govuk-tabs") do
+          within("#cymru-a-lloegr") do
+            expect(page).to have_link("Ychwanegu'r gwyliau banc ar gyfer Cymru a Lloegr at eich calendr (ICS, 14KB)", href: "/gwyliau-banc/cymru-a-lloegr.ics")
+            assert_bank_holiday_table(title: "Gwyliau banc i ddod yng Nghymru a Lloegr", year: "2012", rows: [["Dyddiad", "Diwrnod yr wythnos", "G\u0175yl y banc"], ["25 Rhagfyr", "Dydd Mawrth", "Dydd Nadolig"], ["26 Rhagfyr", "Dydd Mercher", "Dydd San Steffan"]])
```
Done

KludgeKML commented on spec/feature/csv_preview_spec.rb:
```diff
@@ -0,0 +1,243 @@
+  asset_manager_id = 4321
+  asset_media_url_path = "media/#{asset_manager_id}/#{filename}.csv"
+  parent_document_base_path = "/government/important-guidance"
+  parent_document_url = "https://www.test.gov.uk#{parent_document_base_path}"
```
I've made them lets (they're used in multiple places)

KludgeKML commented on spec/feature/csv_preview_spec.rb:
```diff
@@ -0,0 +1,243 @@
+  end
+
+  context "when visiting the preview with special characters in filename" do
+    before do
```
Done

KludgeKML commented on spec/feature/csv_preview_spec.rb:
```diff
@@ -0,0 +1,243 @@
+    csv_file = if malformed
+                 " \"Field 1\"\n"
+               else
+                 nonconvertible ? "\x8D\n" : generate_test_csv(51, 1010)
```
Done

KludgeKML commented on spec/feature/csv_preview_spec.rb:
```diff
@@ -0,0 +1,243 @@
+
+  def setup_content_item(non_legacy_url_path, parent_document_base_path)
+    filename = non_legacy_url_path.split("/").last
+    content_item = { base_path: parent_document_base_path, document_type: "guidance", public_updated_at: "2023-05-27T08:00:07.000+00:00", details: { attachments: [{ title: "Attachment 1", filename: "file.csv", url: "https://www.gov.uk/media/5678/file.csv", file_size: "1024" }, { title: "Attachment 2", filename:, url: "https://www.gov.uk/#{non_legacy_url_path}", file_size: "2048" }] }, links: { or ...
```
Done

KludgeKML commented on spec/requests/csv_preview_spec.rb:
```diff
@@ -0,0 +1,52 @@
+
+  def setup_content_item(non_legacy_url_path, parent_document_base_path)
+    filename = non_legacy_url_path.split "/".last
+    content_item = { base_path: parent_document_base_path, document_type: "guidance", public_updated_at: "2023-05-27T08:00:07.000+00:00", details: { attachments: [{ title: "Attachment 1", filename: "file.csv", url: "https://www.gov.uk/media/5678/file.csv", file_size: "1024" }, { title: "Attachment 2", filename:, url: "https://www.gov.uk/#{non_legacy_url_path}", file_size: "2048" }] }, links: { or ...
```
Done

KludgeKML commented on spec/feature/electoral_look_up_spec.rb:
```diff
@@ -0,0 +1,139 @@
+    visit(electoral_services_path)
+    fill_in("postcode", with: postcode)
+    click_button("Find")
+  end
```
It is used in more contexts than just that one?

KludgeKML commented on spec/feature/find_local_council_spec.rb:
```diff
@@ -0,0 +1,509 @@
+      expect(page.title).to eq("Find your local council - GOV.UK")
+    end
+  end
+  context "when entering a postcode in the search form" do
```
Done

KludgeKML commented on spec/feature/find_local_council_spec.rb:
```diff
@@ -0,0 +1,509 @@
+# the merges and prevents the site breaking during the merge period.
+# See find_local_council_controller#result for more explanation
+
+LOCAL_LINKS_MANAGER_ENDPOINT = Plek.find("local-links-manager")
```
Done

KludgeKML commented on spec/feature/find_local_council_spec.rb:
```diff
@@ -0,0 +1,509 @@
+# where it is useful for a short period of time to mark a county as a unitary if it is
+# becoming one. If we reengineer LLM we can probably remove this, but for now it simplifies
+# the merges and prevents the site breaking during the merge period.
+# See find_local_council_controller#result for more explanation
```
LLM was not reenginered, the test is still valid. Moved the comment.

KludgeKML commented on spec/feature/help_cookies_spec.rb:
```diff
@@ -0,0 +1,24 @@
+RSpec.describe "HelpCookies", type: :feature do
+  context "rendering the cookies setting page" do
+    before do
+      payload = { base_path: "/help/cookies", format: "special_route", title: "Cookies on GOV.UK", description: "You can choose which cookies you're happy for GOV.UK to use." }
```
Done

KludgeKML commented on spec/feature/help_cookies_spec.rb:
```diff
@@ -0,0 +1,24 @@
+    it "renders the cookies setting page correctly" do
+      visit "/help/cookies"
+
+      within("#content") { assert_has_component_title("Cookies on GOV.UK") }
```
Done

KludgeKML commented on spec/feature/homepage_spec.rb:
```diff
@@ -0,0 +1,26 @@
+
+  context "when visiting a Welsh content item first" do
+    before do
+      @payload = { base_path: "/cymraeg", content_id: "d6d6caaf-77db-47e1-8206-30cd4f3d0e3f", document_type: "transaction", locale: "cy", publishing_app: "publisher", rendering_app: "frontend", schema_name: "transaction", title: "Cymraeg", description: "Cynnwys Cymraeg", details: { transaction_start_link: "http://cymraeg.example.com", start_button_text: "Start now" } }
```
Done

KludgeKML commented on spec/feature/error_handling_spec.rb:
```diff
@@ -0,0 +1,54 @@
+
+  context "when the application tries to retrieve an invalid URL from the content store" do
+    it "returns 404 status" do
+      content_store_throws_exception_for("/foo", GdsApi::InvalidUrl)
```
Hmm. Maybe? I don't know if it's used anywhere except Frontend.

KludgeKML commented on spec/feature/account_home_spec.rb:
```diff
@@ -0,0 +1,20 @@
+    stub_content_store_has_item("/")
+  end
+
+  context "/account/home" do
```
Yes, we're using the default rather than setting it explicitly (no need to do that), and in a system test we can just test if we've been bounced to the new URL (in which case a redirect has happened)

KludgeKML commented on spec/functional/simple_smart_answers_controller_spec.rb:
```diff
@@ -0,0 +1,110 @@
+RSpec.describe SimpleSmartAnswersController, type: :controller do
+  include GovukAbTesting::RspecHelpers
```
Done, thanks!

KludgeKML commented on spec/functional/transaction_controller_spec.rb:
```diff
@@ -0,0 +1,82 @@
+RSpec.describe TransactionController, type: :controller do
+  include GovukAbTesting::RspecHelpers
```
Done!

KludgeKML commented on spec/functional/travel_advice_controller_spec.rb:
```diff
@@ -0,0 +1,49 @@
+require "gds_api/test_helpers/content_store"
+
+RSpec.describe TravelAdviceController, type: :controller do
+  include GdsApi::TestHelpers::ContentStore
```
Nope! Thanks

KludgeKML commented on spec/unit/helpers/application_helper_spec.rb:
```diff
@@ -0,0 +1,53 @@
+RSpec.describe ApplicationHelper, type: :view do
+  include ContentStoreHelpers
+
+  tests(ApplicationHelper)
```
Done!

KludgeKML commented on spec/unit/helpers/currency_helper_spec.rb:
```diff
@@ -0,0 +1,15 @@
+RSpec.describe CurrencyHelper, type: :view do
+  include(CurrencyHelper)
```
Done!

KludgeKML commented on spec/unit/helpers/phone_number_helper_spec.rb:
```diff
@@ -0,0 +1,17 @@
+RSpec.describe PhoneNumberHelper, type: :view do
+  include(PhoneNumberHelper)
```
Done!

KludgeKML commented on spec/unit/models/uprn_spec.rb:
```diff
@@ -0,0 +1,46 @@
+RSpec.describe Uprn, type: :model do
```
I think this is fixed, although it looks a bit spaced out because of the "space between test setup and expect" I've been doing, which in this simple file with lots of 2-line tests makes it look very whitespacey.

KludgeKML commented on spec/unit/presenters/calendar_content_item_spec.rb:
```diff
@@ -0,0 +1,38 @@
+
+  def content_item(slug: nil)
+    calendar = Calendar.find("bank-holidays")
+    CalendarContentItem.new(calendar, slug:)
```
Done!

KludgeKML commented on spec/unit/presenters/content_item_presenter_spec.rb:
```diff
@@ -0,0 +1,52 @@
+RSpec.describe ContentItemPresenter, type: :model do
+  def subject(content_item)
+    ContentItemPresenter.new(content_item.deep_stringify_keys!)
+  end
```
Done!

KludgeKML commented on spec/unit/presenters/licence_details_presenter_spec.rb:
```diff
@@ -0,0 +1,158 @@
+      @presented_the_other_licence_authority = { "name" => "The Other Licence Authority", "slug" => "the-other-licence-authority", "contact" => nil, "actions" => { "apply" => [{ "url" => "the-other-licence-authority/apply-1", "introduction" => nil, "description" => nil, "payment" => nil, "uses_licensify" => true, "uses_authority_url" => true }] } }
+    end
+
+    context "#authority" do
```
Done!

KludgeKML replied:
These are confusions due to the commit that fixed all these things being merged into the following commit not this one. I've fixed that now, so will mark all these as resolved.

KludgeKML commented on spec/unit/presenters/local_authority_presenter_spec.rb:
```diff
@@ -0,0 +1,39 @@
+    before do
+      @local_authority_payload = { "name" => "Westminster City Council", "snac" => "00BK", "gss" => "E060000034", "tier" => "district", "homepage_url" => "http://westminster.example.com/" }
+      @local_authority_presenter = LocalAuthorityPresenter.new(@local_authority_payload)
+    end
```
Done!

KludgeKML commented on spec/feature/find_local_council_spec.rb:
```diff
@@ -0,0 +1,509 @@
+
+  context "when visiting the start page" do
+    before do
+      allow(Frontend).to receive(:govuk_website_root).and_return("https://www.test.gov.uk")
```
Turns out it doesn't. Removed.

KludgeKML committed:
Fix roadmap meta tag

leenagupte reviewed (commented):
I got as far as reviewing commit bd648fb5bf625bf90d5560622c004b28d2c6ad1b "Convert travel_advice_controller functional tests"

It's looking really good. Thank you for unflattening the hashes. There's so much happening, but the changes are well laid out and easy to follow, even though there are so many!

There are a few things that I've spotted:

1. The conversion didn't work correctly for some of the `not_to` expectations in integration tests. I've commented on the ones I've spotted.
2. In the calendars request tests, and perhaps some others, bad routes are being tested that could do with being separated out as they're testing routing and not calendars.
3. This is the big one. The filenames and "describe" method names all contain the word "controller". I think these should be removed / replaced. I've commented on the first few tests, but it applies to all of them.

I'll get back to reviewing this tomorrow.

leenagupte reviewed (commented):
I've finished reviewing the helper tests.

I think they need an update to the `type` otherwise they look good.

leenagupte reviewed (commented):
I've finished reviewing the model and presenter tests, as far as [Convert travel_advice_index presenter unit](tests3785aab2ec2f52caff7b76614ee6624173685a90)

A couple of observations:
1. `described_class.new` should be used consistently rather than using the full name of the class
2. All of the presenter specs have a `type` of `model` which should be removed. 

Apart from that I think they're looking good.

KludgeKML commented on Rakefile:
```diff
@@ -3,20 +3,15 @@
-
 Rake::Task[:default].clear if Rake::Task.task_defined?(:default)
-task default: %i[test_unit jasmine pact:verify lint]
+task default: %i[lint spec jasmine pact:verify]
```
Yeah, I hate that lint surprise. (marking as resolved)

KludgeKML committed:
Update spec_helper

- Remove autogenerated code (we want all the defaults
  anyway), and replace with the rails loading
  code (We probably don't run individual tests enough
  to justify keeping spec_helper lightweight)
- Move govukTest config support into spec_helper
  from separate support file.
- Add ** glob to allow subdirectories of support for
  custom matchers
- Add infer_spec_type_from_file_location! to allow
  non-explicit spec types.

leenagupte commented on spec/unit/format_routing_constraint_spec.rb:
```diff
@@ -1,56 +1,58 @@
-require "test_helper"
+RSpec.describe FormatRoutingConstraint, type: :model do
```
`type` can be removed here.

KludgeKML replied:
Done

leenagupte commented on spec/unit/format_routing_constraint_spec.rb:
```diff
@@ -1,56 +1,58 @@
-      should "return true if format matches" do
-        assert subject(@format).matches?(@request)
+      it "returns true if format matches" do
+        expect(subject(@format).matches?(@request)).to be true
```
Should probably use `described_class` here rather than `subject`.

KludgeKML replied:
Have reworked this file to use described_class and let a bit more

leenagupte reviewed (commented):
I've finished reviewing!

Nearly everything is a comment on formatting, and again a few places where the `false` logic is a bit strange.

This is a mammoth PR! Well done for persevering with it.

I'll be on leave, but I'm happy for @hannako to approve once the formatting has been fixed.

KludgeKML committed:
Convert homepage_controller functional tests

KludgeKML commented on spec/components/all_components_spec.rb:
```diff
@@ -0,0 +1,48 @@
+RSpec.describe "AllComponents", type: :controller do
+  Dir.glob("app/views/components/*.erb").each do |filename|
+    template = filename.split("/").last
+    component_name = template.sub("_", "").sub(".html", "").sub(".erb", "").gsub("-", "_")
```
It's because the component names are loaded from their partials, so the sub removes the first _, then the gsub at the end converts the rest of the name from kebab case to snake case. I think for the moment we should keep this (because it matches the corresponding code in govuk_publishing_components), but we should look later at making all of these shared examples that are loaded directly from the components gem.

KludgeKML commented on spec/components/all_components_spec.rb:
```diff
@@ -0,0 +1,48 @@
+      end
+
+      it "has the correct documentation" do
+        yaml = YAML.safe_load_file(yaml_file, permitted_classes: [Time])
```
(Positive, marking as resolved). :)

KludgeKML commented on app/models/calendar.rb:
```diff
@@ -1,5 +1,5 @@
 class Calendar
-  REPOSITORY_PATH = Rails.env.test? ? "test/fixtures" : "lib/data"
+  REPOSITORY_PATH = "lib/data".freeze
```
Positive, marking resolved

KludgeKML commented on spec/system/csv_preview_spec.rb:
```diff
@@ -0,0 +1,282 @@
+RSpec.describe "CsvPreview", type: :system do
+  include GdsApi::TestHelpers::AssetManager
+
+  let(:filename) { "filename" }
+  let(:asset_manager_id) { 4321 }
+  let(:asset_media_url_path) { "media/#{asset_manager_id}/#{filename}.csv" }
+  let(:parent_document_base_path) { "/government/important-guidance" }
+  let(:parent_document_url) { "https://www.test.gov.uk#{parent_document_base_path}" }
```
(positive, marking resolved)

KludgeKML commented on spec/system/csv_preview_spec.rb:
```diff
@@ -0,0 +1,282 @@
+    csv_file = if malformed
+                 " \"Field 1\"\n"
+               elsif nonconvertible
+                 "\x8D\n"
```
Positive, marking resolved

KludgeKML commented on spec/system/local_transactions_spec.rb:
```diff
@@ -0,0 +1,519 @@
+
+  before do
+    configure_locations_api_and_local_authority("SW1A 1AA", %w[westminster], 5990)
+    @payload = {
```
It was! But it's done now, and all better for it. (positive, marking resolved)

KludgeKML commented on spec/system/place_spec.rb:
```diff
@@ -0,0 +1,299 @@
+    ]
+  end
+
+  context "when visiting the start page" do
```
Added

KludgeKML commented on spec/routing/simple_smart_answers_routing_spec.rb:
```diff
@@ -0,0 +1,31 @@
+RSpec.describe "SimpleSmartAnswersRouting", type: :routing do
```
(Positive, marked resolved)

KludgeKML commented on spec/routing/simple_smart_answers_routing_spec.rb:
```diff
@@ -0,0 +1,31 @@
+
+  context "for the start page" do
+    it "routes the start page to the SimpleSmartAnswer controller" do
+      assert_recognizes({ controller: "simple_smart_answers", action: "show", slug: "fooey" }, "/fooey")
```
Ah yes, it's exactly the same. Fixed.

KludgeKML commented on spec/unit/helpers/application_helper_spec.rb:
```diff
@@ -0,0 +1,51 @@
+RSpec.describe ApplicationHelper, type: :view do
```
I actually think probably no type is okay here, but what I've done is made them helper tests by removing all the explicit test types and just rearranging into directories so that `config.infer_spec_type_from_file_location!` works (I've done that for all types - system and requests were already like this, but it means moving helpers and models out into their own directories).

KludgeKML commented on spec/unit/helpers/currency_helper_spec.rb:
```diff
@@ -0,0 +1,15 @@
+RSpec.describe CurrencyHelper, type: :view do
```
(See previous comment)

KludgeKML commented on spec/unit/helpers/recruitment_banner_helper_spec.rb:
```diff
@@ -0,0 +1,44 @@
+  end
+
+  it "recruitment_banner returns banners that include the current url" do
+    actual_banners = recruitment_banner
```
Yeah, it took me a moment to work out what this was. I've called it directly and wrapped this in a describe block with the method (recruitment_banner)'s name.

KludgeKML commented on spec/unit/helpers/recruitment_banner_helper_spec.rb:
```diff
@@ -0,0 +1,44 @@
+    if @recruitment_banners_data["banners"].present?
+      recruitment_banners.each do |banner|
+        expect(banner.key?("suggestion_text")).to be true
+        expect(banner["suggestion_text"].blank?).to be true
```
Yeah, agreed. Done.

KludgeKML commented on spec/unit/helpers/recruitment_banner_helper_spec.rb:
```diff
@@ -0,0 +1,44 @@
+        expect(banner.key?("suggestion_text")).to be true
+        expect(banner["suggestion_text"].blank?).to be true
+        expect(banner.key?("suggestion_link_text")).to be true
+        expect(banner["suggestion_link_text"].blank?).to be true
```
Done

KludgeKML commented on spec/unit/helpers/recruitment_banner_helper_spec.rb:
```diff
@@ -0,0 +1,44 @@
+        expect(banner.key?("suggestion_link_text")).to be true
+        expect(banner["suggestion_link_text"].blank?).to be true
+        expect(banner.key?("survey_url")).to be true
+        expect(banner["survey_url"].blank?).to be true
```
Done

KludgeKML commented on spec/unit/helpers/recruitment_banner_helper_spec.rb:
```diff
@@ -0,0 +1,44 @@
+        expect(banner.key?("survey_url")).to be true
+        expect(banner["survey_url"].blank?).to be true
+        expect(banner.key?("page_paths")).to be true
+        expect(banner["page_paths"].blank?).to be true
```
Done

KludgeKML commented on spec/unit/services/calendar_publisher_spec.rb:
```diff
@@ -0,0 +1,40 @@
+    expect(publishing_api).to receive(:patch_links).with("58f79dbd-e57f-4ab2-ae96-96df5767d1b2", { links: ({ primary_publishing_organisation: %w[af07d5a5-df63-4ddc-9383-6a666845ebe9] }) })
+
+    calendar = Calendar.find("bank-holidays")
+    CalendarPublisher.new(calendar).publish
```
Done

KludgeKML replied:
Done

KludgeKML commented on spec/unit/simple_smart_answers/flow_spec.rb:
```diff
@@ -0,0 +1,141 @@
+module SimpleSmartAnswers
+  RSpec.describe Flow, type: :model do
```
Removed

KludgeKML commented on spec/unit/api_error_routing_constraint_spec.rb:
```diff
@@ -0,0 +1,21 @@
+require "api_error_routing_constraint"
+
+RSpec.describe ApiErrorRoutingConstraint, type: :model do
```
Removed from all.

KludgeKML commented on spec/unit/bank_holiday_generator_spec.rb:
```diff
@@ -0,0 +1,86 @@
+    end
+  end
+
+  nations = %w[england-and-wales scotland northern-ireland]
```
It's being used 2 lines down, and only there.. I've simplified by removing the temp vars.

KludgeKML committed:
Convert licence_transaction presenter unit tests

KludgeKML commented on spec/system/simple_smart_answers_spec.rb:
```diff
@@ -0,0 +1,489 @@
+      expect(page).to have_selector("title", exact_text: "What...is your name? - The Bridge of Death - GOV.UK", visible: :all)
+      expect(page).to have_selector("meta[name=robots][content=noindex]", visible: :all)
+    end
+    within("#content") do
```
Added

KludgeKML commented on spec/unit/presenters/licence_details_presenter_spec.rb:
```diff
@@ -1,18 +1,136 @@
-    @licence_authority_with_no_actions = { "authorityName" => "The Licence Authority with no actions", "authoritySlug" => "the-licence-authority-with-no-actions", "authorityInteractions" => {} }
-    @licence_authority_licence = { "isLocationSpecific" => false, "isOfferedByCounty" => false, "geographicalAvailability" => %w[England Wales], "issuingAuthorities" => [@the_one_licence_authority] }
-    @multiple_authorities_and_location_specific_licence = { "isLocationSpecific" => true, "isOfferedByCounty" => false, "geographicalAvailability" => %w[England Wales], "issuingAuthorities" => [@the_other_licence_authority, @the_one_licence_authority] }
+    @local_authority_licence = {
```
A fix got squashed into the wrong commit. It's in the right one now.

KludgeKML committed:
Convert currency_helper unit tests

KludgeKML committed:
Convert all simple_smart_answer unit tests

KludgeKML commented on spec/unit/presenters/licence_details_presenter_spec.rb:
```diff
@@ -0,0 +1,158 @@
+    @multiple_authorities_and_location_specific_licence = { "isLocationSpecific" => true, "isOfferedByCounty" => false, "geographicalAvailability" => %w[England Wales], "issuingAuthorities" => [@the_other_licence_authority, @the_one_licence_authority] }
+  end
+
+  context "#single_licence_authority_present?" do
```
These are confusions due to the commit that fixed all these things being merged into the following commit not this one. I've fixed that now, so will mark all these as resolved.

KludgeKML commented on spec/system/find_local_council_spec.rb:
```diff
@@ -0,0 +1,443 @@
+        it "includes the search for another postcode link" do
+          expect(page).to have_link("Search for a different postcode", href: "/find-local-council", exact: true)
+        end
+        it "includes the search result text in the page title" do
```
Done

KludgeKML commented on spec/system/find_local_council_spec.rb:
```diff
@@ -0,0 +1,443 @@
+          expect(page).to have_link("Search for a different postcode", href: "/find-local-council", exact: true)
+        end
+        it "includes the search result text in the page title" do
+          expect(page).to have_title("Find your local council: #{I18n.t('formats.local_transaction.search_result')} - GOV.UK", exact: true)
```
Positive, marking resolved.

KludgeKML commented on spec/system/find_local_council_spec.rb:
```diff
@@ -0,0 +1,443 @@
+  end
+end
+
+def stub_local_links_manager_has_a_district_and_unitary_local_authority(district_slug, unitary_slug, district_snac: "00AG", unitary_snac: "00LC", local_custodian_code: nil)
```
No, it's subtly different (this is mimicking the unlikely but possible situation in which we have a district but the parent is a unitary rather than county LA. There isn't a test stub for that and it's only relevant here, so... seems reasonable to add a new method)

KludgeKML commented on spec/support/button_as_link_helpers.rb:
```diff
@@ -0,0 +1,23 @@
+  def process_button_attributes(attrs)
+    match_assert = ""
+    # match_assert = ".gem-c-button"
+    # match_assert << ".gem-c-button--start" if attrs[:start]
```
No, it should be deleted. Done

KludgeKML commented on spec/system/local_transactions_spec.rb:
```diff
@@ -0,0 +1,519 @@
+      end
+
+      it "does not show the transaction information" do
+        expect(page.has_content?("owning or looking after a bear")).to be false
```
Done

KludgeKML commented on spec/system/local_transactions_spec.rb:
```diff
@@ -0,0 +1,519 @@
+      end
+
+      it "does not show a postcode error" do
+        expect(page.has_selector?(".location_error")).to be false
```
Done

KludgeKML commented on spec/system/local_transactions_spec.rb:
```diff
@@ -0,0 +1,519 @@
+      end
+
+      it "does not show link to change location" do
+        expect(page.has_link?("(change location)")).to be false
```
Done

KludgeKML commented on spec/system/simple_smart_answers_spec.rb:
```diff
@@ -0,0 +1,489 @@
+      within(".govuk-fieldset__legend") do
+        expect(page).to have_content("2. What...is your favorite colour?")
+      end
+      expect(page.has_selector?(".govuk-error-message", text: "Please answer this question")).to be false
```
Done

KludgeKML commented on spec/system/travel_advice_spec.rb:
```diff
@@ -0,0 +1,86 @@
+RSpec.describe "TravelAdvice", type: :system do
+  # Necessary because Capybara's has_content? method normalizes spaces in the document
```
Nope, removed.

KludgeKML commented on spec/system/travel_advice_spec.rb:
```diff
@@ -0,0 +1,86 @@
+  # Necessary because Capybara's has_content? method normalizes spaces in the document
+  # However, it doesn't normalize spaces in the query string, so if you're looking for a string
+  # with 2 spaces in it (e.g. when the date is a single digit with a space prefix), it will fail.
+  def de_dup_spaces(string)
```
Removed

KludgeKML commented on spec/requests/calendar_controller_spec.rb:
```diff
@@ -0,0 +1,159 @@
+RSpec.describe CalendarController, type: :request do
```
Done

KludgeKML commented on spec/requests/electoral_controller_spec.rb:
```diff
@@ -0,0 +1,78 @@
+RSpec.describe ElectoralController, type: :request do
```
Done

KludgeKML commented on spec/requests/place_controller_spec.rb:
```diff
@@ -0,0 +1,72 @@
+  include GdsApi::TestHelpers::PlacesManager
+
+  valid_postcode = "SW1A 2AA"
+  invalid_postcode = "1234 2AA"
```
Done

KludgeKML commented on spec/requests/place_controller_spec.rb:
```diff
@@ -0,0 +1,72 @@
+    stub_places_manager_places_request("slug", query_hash, return_data, 400)
+  end
+
+  context "GET show" do
```
Done

KludgeKML replied:
(Have renamed this and the one below to what they're getting and posting)

KludgeKML commented on spec/requests/place_controller_spec.rb:
```diff
@@ -0,0 +1,72 @@
+    end
+  end
+
+  context "POST find" do
```
Done (Have renamed this and the one above to what they're getting and posting)

KludgeKML commented on spec/requests/random_controller_spec.rb:
```diff
@@ -0,0 +1,59 @@
+RSpec.describe RandomController, type: :request do
+  context "random page route" do
```
Removed

KludgeKML commented on spec/requests/simple_smart_answers_controller_spec.rb:
```diff
@@ -0,0 +1,192 @@
+    }
+  end
+
+  context "GET show" do
```
Renamed to what it's getting

KludgeKML commented on spec/requests/simple_smart_answers_controller_spec.rb:
```diff
@@ -0,0 +1,192 @@
+      content_store_has_random_item(base_path: "/the-bridge-of-death", schema: "simple_smart_answer")
+    end
+
+    context "for live content" do
```
Removed these contexts as they don't add anything.

KludgeKML commented on spec/requests/simple_smart_answers_controller_spec.rb:
```diff
@@ -0,0 +1,192 @@
+  end
+
+  context "GET 'flow'" do
+    context "for a simple_smart_answer slug" do
```
Done

KludgeKML commented on spec/unit/models/location_error_spec.rb:
```diff
@@ -0,0 +1,30 @@
+  describe "#initialize" do
+    context "when given a postcode error with a message and sub message" do
+      it "sets the message and the sub message" do
+        error = LocationError.new("fullPostcodeNoLocationsApiMatch")
```
Done

KludgeKML commented on spec/unit/models/location_error_spec.rb:
```diff
@@ -0,0 +1,30 @@
+
+    context "when given a postcode error without a message" do
+      it "sets default message and sub message" do
+        error = LocationError.new("some_error")
```
Done

KludgeKML commented on spec/unit/models/uprn_spec.rb:
```diff
@@ -0,0 +1,55 @@
+    end
+  end
+
+  describe "#errors" do
```
Done

KludgeKML commented on spec/unit/presenters/calendar_content_item_spec.rb:
```diff
@@ -0,0 +1,38 @@
+require "govuk_schemas/assert_matchers"
+
+RSpec.describe CalendarContentItem, type: :model do
```
Done (as mentioned in updated description, have removed all but a few explicit types and moved to relying on the infer_type_from_file_location! config option)

KludgeKML commented on spec/unit/presenters/content_item_presenter_spec.rb:
```diff
@@ -0,0 +1,52 @@
+RSpec.describe ContentItemPresenter, type: :model do
```
Done

KludgeKML commented on spec/unit/presenters/electoral_presenter_spec.rb:
```diff
@@ -0,0 +1,75 @@
+
+  def electoral_presenter(payload)
+    ElectoralPresenter.new(payload)
+  end
```
Done

KludgeKML commented on spec/unit/presenters/licence_details_presenter_spec.rb:
```diff
@@ -0,0 +1,158 @@
+    @licence_authority_with_no_actions = { "authorityName" => "The Licence Authority with no actions", "authoritySlug" => "the-licence-authority-with-no-actions", "authorityInteractions" => {} }
+    @licence_authority_licence = { "isLocationSpecific" => false, "isOfferedByCounty" => false, "geographicalAvailability" => %w[England Wales], "issuingAuthorities" => [@the_one_licence_authority] }
+    @multiple_authorities_and_location_specific_licence = { "isLocationSpecific" => true, "isOfferedByCounty" => false, "geographicalAvailability" => %w[England Wales], "issuingAuthorities" => [@the_other_licence_authority, @the_one_licence_authority] }
+  end
```
Done (this is from the squash problem, marking resolved)

KludgeKML commented on spec/unit/presenters/licence_details_presenter_spec.rb:
```diff
@@ -0,0 +1,158 @@
+  context "authority look up methods" do
+    before do
+      @presented_the_one_licence_authority = { "name" => "The One Licence Authority", "slug" => "the-one-licence-authority", "contact" => nil, "actions" => { "apply" => [{ "url" => "the-one-licence-authority/apply-1", "introduction" => nil, "description" => nil, "payment" => nil, "uses_licensify" => true, "uses_authority_url" => true }] } }
+      @presented_the_other_licence_authority = { "name" => "The Other Licence Authority", "slug" => "the-other-licence-authority", "contact" => nil, "actions" => { "apply" => [{ "url" => "the-other-licence-authority/apply-1", "introduction" => nil, "description" => nil, "payment" => nil, "uses_licensify" => true, "uses_authority_url" => true }] } }
```
Done (this is from the squash problem, marking resolved)

KludgeKML commented on spec/unit/presenters/licence_details_presenter_spec.rb:
```diff
@@ -0,0 +1,158 @@
+        it "returns nil if no matched authority found" do
+          subject = LicenceDetailsPresenter.new(@multiple_authorities_and_location_specific_licence, "a-funky-licence-authority")
+
+          expect(subject.authority).to(be_nil)
```
Done

KludgeKML commented on spec/unit/presenters/licence_details_presenter_spec.rb:
```diff
@@ -0,0 +1,158 @@
+        it "returns nil" do
+          subject = LicenceDetailsPresenter.new(@licence_authority_licence.merge("issuingAuthorities" => []))
+
+          expect(subject.authority).to(be_nil)
```
Done

KludgeKML commented on spec/unit/presenters/licence_details_presenter_spec.rb:
```diff
@@ -0,0 +1,158 @@
+    it "raises RecordNotFound if not available" do
+      subject = LicenceDetailsPresenter.new(@local_authority_licence, nil, "complain")
+
+      expect { subject.action }.to(raise_error(RecordNotFound))
```
Done

KludgeKML commented on spec/unit/presenters/local_authority_presenter_spec.rb:
```diff
@@ -0,0 +1,37 @@
+      "homepage_url" => "http://westminster.example.com/",
+    }
+  end
+  let(:local_authority_presenter) { LocalAuthorityPresenter.new(local_authority_payload) }
```
Done

KludgeKML commented on spec/unit/presenters/place_presenter_spec.rb:
```diff
@@ -0,0 +1,38 @@
+RSpec.describe PlacePresenter, type: :model do
+  def subject(content_item)
+    @subject ||= PlacePresenter.new(content_item.deep_stringify_keys!)
```
Yeah, weird. Removed

KludgeKML commented on spec/unit/services/calendar_publisher_spec.rb:
```diff
@@ -0,0 +1,40 @@
+RSpec.describe CalendarPublisher, type: :model do
```
Done

KludgeKML commented on spec/unit/services/csv_preview_service_spec.rb:
```diff
@@ -0,0 +1,144 @@
+RSpec.describe CsvPreviewService, type: :model do
```
Done

KludgeKML commented on spec/system/csv_preview_spec.rb:
```diff
@@ -0,0 +1,282 @@
+    end
+
+    it "redirect to parent" do
+      get @preview_path
```
So! Basically, I think the solution here is to make this a request test. It's only really interested in the redirection, and because the scaffolding we'd need to set this up as a system test is quite awkward, it's easier to split it out as a request test (which is clearly what it was with the get method) and create an AssetManagerHelper so that we don't repeat the complex setup methods. I've done that now.

hannako replied:
Request spec is a good idea 👍 

But you've still got the instance variable storing the path. I looked at all the other request specs in this PR and there's only [one other place](https://github.com/alphagov/frontend/pull/4090/files#diff-cb0b18e6af1af6949d2ad7b83c52dcade8ed005d1cc86057e536fdd7e3f70297) we do `get @path`, everywhere else we're doing `get "/path-blah"`

So I guess the easiest fix would just be to do `get "/#{path_from_filename[asset_manager_filename]}/preview"`  but that's a bit of a confusing path to wrap your head around. Is this any better:

```
it "redirects to parent" do
  preview_path = "/#{path_from_filename[asset_manager_filename]}/preview"
  get preview_path

  expect(response).to redirect_to(parent_document_url)
end
```
Don't think it's a big deal though, so feel free to leave it as is if you prefer.

KludgeKML replied:
Ah, the problem with that is that path_from_filename is a proc created in the before block, which isn't available in the it block. But if I make that a method (which is a bit less surprising anyway, then this works nicely.

KludgeKML commented on spec/system/sessions_spec.rb:
```diff
@@ -0,0 +1,47 @@
+      @redirect_path = "/email/subscriptions/account/confirm?frequency=immediately&return_to_url=true&topic_id=some-page-with-notifications"
+      given_a_successful_login_attempt
+
+      expect_to_be_redirected { when_i_return_from_digital_identity }
```
Yeah, this is a bit of a crazy one. I don't fully understand why it needs this bizarre yield/rescue block or what that's supposed to be doing. If it's okay, let's come back to this in a future PR.

hannako replied:
I think it might be worth adding back in the original comment that's presumably been lost with the conversion? As it kind of makes more sense: https://github.com/alphagov/frontend/blob/94fe22def95df71cf6f89943a86c35d1e4fb7f76/test/integration/sessions_test.rb#L52-L58

KludgeKML replied:
Okay, playing around with this a bit I think I've got a better plan (and I've commented it), which is to use the normal `expect(page.current_url).to eq (redirect path)`, and to make that route available by temporarily adding it pointing to the homepage.

KludgeKML commented on spec/requests/calendar_controller_spec.rb:
```diff
@@ -0,0 +1,159 @@
+      expect(response).to have_http_status(:not_found)
+    end
+
+    it "does not route an invalid slug format, and does not try to look up the calendar" do
```
This one definitely is a routing test, have moved it there.

KludgeKML commented on spec/requests/calendar_controller_spec.rb:
```diff
@@ -0,0 +1,159 @@
+      end
+    end
+
+    it "returns 404 for a non-existent calendar" do
```
It's tricky, this is testing both expected behaviour (if we have a content item that says it's a calendar but we don't know about the calendar do we crash or handle it gracefully) and confirming the path that handles that behaviour is what we expect (which is more like a routing test). I've split it out into both - this test remains to check the 404, but there's also a routing test which checks that the right route is handling it. Without the routing test we could be getting a 404 from anywhere, without the request test we wouldn't know if we were doing the right thing.

hannako replied:
I agree with this approach 👍

hannako reviewed (approved):
END OF AN ERA!

Well done on not giving up in the face of 8 thousand review comments 💪

KludgeKML commented:
Those 8 thousand review comments helped it move from a poorly formed mass of clay into a fine china pot (with a few blemishes). Thanks @hannako and @leenagupte for all the suggestions!
